File: src/main/java/com/tjh/riskfactor/entity/form/Task.java
package com.tjh.riskfactor.entity.form;

import lombok.Data;
import lombok.experimental.Accessors;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonProperty;
import static com.fasterxml.jackson.annotation.JsonFormat.Shape;

import com.tjh.riskfactor.entity.Group;

import javax.persistence.*;
import java.util.Date;
import java.util.List;

@Data @Entity
@Table(name = "task")
@Accessors(chain = true)
public class Task {

    @PrePersist
    private void assignFields() {
        if(mtime == null)
            mtime = new Date();
    }

    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Integer id;

    @ManyToOne
    @JoinColumn(name = "gid", referencedColumnName = "id", nullable = false)
    @JsonIgnore
    private Group group;

    @Column(nullable = false)
    private String name;

    @Column(nullable = false)
    @JsonFormat(shape = Shape.STRING, pattern = "yyyy-MM-dd")
    private Date mtime;

    @OneToMany(fetch = FetchType.EAGER, cascade = CascadeType.REMOVE)
    @JoinTable(name = "task_sections",
        joinColumns = @JoinColumn(name = "tid"),
        inverseJoinColumns = @JoinColumn(name = "sid")
    )
    @OrderColumn(name = "seq", nullable = false)
    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    private List<Section> sections;

    // 用来给JSON用的，传入内容为用户组名
    @JsonProperty(access = JsonProperty.Access.WRITE_ONLY)
    @Transient private String center;

    @JsonProperty(value = "center", access = JsonProperty.Access.READ_ONLY)
    @Transient public String getCenterName() {
        return group.getDisplayName();
    }

}

File: backend/src/main/kotlin/com/tjh/riskfactor/entity/form/Task.kt
package com.tjh.riskfactor.entity.form

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonFormat
import com.fasterxml.jackson.annotation.JsonProperty

import com.tjh.riskfactor.entity.IEntity
import com.tjh.riskfactor.entity.Group

import javax.persistence.*
import java.util.Date

/**
 * 一个项目
 */
@Entity @Table(name = "task")
class Task(
    @get:Column(nullable = false)
    var name: String
): IEntity() {
    @get:Column(nullable = false)
    @get:JsonFormat(shape = JsonFormat.Shape.STRING, pattern = "yyyy-MM-dd")
    var mtime: Date = Date()

    @get:ManyToOne
    @get:JoinColumn(name = "gid", referencedColumnName = "id", nullable = false)
    @JsonIgnore
    lateinit var group: Group

    @get:OneToMany(cascade = [CascadeType.REMOVE])
    @get:JoinTable(name = "task_sections",
        joinColumns = [JoinColumn(name = "tid")],
        inverseJoinColumns = [JoinColumn(name = "sid")]
    )
    @get:OrderColumn(name = "seq", nullable = false)
    @set:JsonProperty
    var sections: MutableList<Section> = mutableListOf()

    // 用来给JSON用的，传入内容为用户组名
    @set:JsonProperty
    @get:Transient var center: String = ""

    @JsonProperty(value = "center", access = JsonProperty.Access.READ_ONLY)
    @Transient fun getCenterName() = group.displayName
}
